casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala

```diff
@@ -0,0 +1,161 @@
+package coop.rchain.casper.api
+
+import cats.effect.Sync
+import cats.implicits._
+import com.google.protobuf.ByteString
+import coop.rchain.casper.engine.Engine
+import coop.rchain.casper.{EngineWithCasper, SafetyOracle}
+import coop.rchain.casper.helper._
+import coop.rchain.casper.helper.BlockGenerator._
+import coop.rchain.casper.protocol._
+import coop.rchain.casper.util.GenesisBuilder._
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.casper.util.ConstructDeploy.basicDeployData
+import coop.rchain.crypto.signatures.Secp256k1
+import coop.rchain.metrics.Metrics
+import coop.rchain.shared.Cell
+import coop.rchain.shared.scalatestcontrib.effectTest
+import monix.eval.Task
+import org.scalatest.{EitherValues, FlatSpec, Matchers}
+import monix.execution.Scheduler.Implicits.global
+
+import scala.collection.immutable.Map
+
+class LastFinalizedAPITest
+    extends FlatSpec
+    with Matchers
+    with EitherValues
+    with BlockGenerator
+    with BlockDagStorageFixture {
+  val validatorKeyPairs = (1 to 3).map(_ => Secp256k1.newKeyPair)
+  val (_, validatorPks) = validatorKeyPairs.unzip
+  val validators        = validatorPks.map(pk => ByteString.copyFrom(pk.bytes))
+  val v1                = validators(0)
+  val v2                = validators(1)
+  val v3                = validators(2)
+  val genesisParameters = buildGenesisParameters(
+    validatorKeyPairs,
+    Map(validatorPks(0) -> 50, validatorPks(1) -> 45, validatorPks(2) -> 5)
+  )
+  val genesisContext = buildGenesis(genesisParameters)
+
+  implicit val metricsEff = new Metrics.MetricsNOP[Task]
+
+  /*
+   * DAG Looks like this:
+   *
+   *           b5
+   *           |
+   *           b4
+   *           |
+   *           b3
+   *           |
+   *           b2 <- last finalized block
+   *           |
+   *           b1
+   *           |
+   *         genesis
+   */
+  "isFinalized" should "return true for ancestors of last finalized block" in effectTest {
+    HashSetCasperTestNode.networkEff(genesisContext, networkSize = 3).use {
+      case n1 +: n2 +: n3 +: _ =>
+        import n1.{blockStore, cliqueOracleEffect, logEff}
+        val engine = new EngineWithCasper[Task](n1.casperEff)
+        for {
+          produceDeploys <- (0 until 5).toList.traverse(i => basicDeployData[Task](i))
+
+          b1 <- n1.addBlock(produceDeploys(0))
+          _  <- n2.receive()
+          _  <- n3.receive()
```

Comment:
Thx for rephrasing the tests! :)

Could you add a `.publishBlock(deployDatums: DeployData*)(nodes: HashSetCasperTestNode: _*)` method to HashSetCasperTestNode, that would do `nodes.filter(_ != this).traverse_(_.receive())`? This way we could

`b1 <- n1.publishBlock((produceDeploys(0))(nodes), where `val nodes = List(n1, n2, n3)`

Reply:
Done